net/tstun/mtu.go

```diff
@@ -1,33 +1,171 @@
+// etc.). This MTU can change at any time. Setting the MTU this way goes through
+// the MTU() method of tailscale's TUN wrapper.
+//
+// Maximum probed MTU: This is the largest MTU size that we send probe packets
+// for.
+//
+// Safe MTU: If the tailscale TUN MTU is set to this value, almost all packets
+// will get to their destination. Tailscale defaults to this MTU in the absence
+// of path MTU probe information or user MTU configuration. We may occasionally
+// find a path that needs a smaller MTU but it is very rare.
+//
+// Peer MTU: This is the path MTU to a peer's current best endpoint. It defaults
+// to the Safe MTU unless we have path MTU probe results that tell us otherwise.
+//
+// Initial MTU: This is the MTU tailscaled creates the TUN with. In order of
+// priority, it is:
+//
+// 1. If set, the value of TS_DEBUG_MTU clamped to a maximum of 65536
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg
+//    overhead
+// 3. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+//
+// Current MTU: This the MTU of the tailscale TUN at any given moment
+// after TUN creation. In order of priority, it is:
+//
+// 1. The MTU set by the user via the OS, if it has ever been set
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg
+//    overhead
+// 4. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+
+// TunMTU is used to describe MTUs for the tailscale TUN.
+type TunMTU uint32
+
+// WireMTU is used to describe MTUs for the underlying network devices.
+type WireMTU uint32
 
 const (
-	maxMTU     uint32 = 65536
-	defaultMTU uint32 = 1280
+	// Largest MTU we will consider for the Tailscale TUN. This is inherited
+	// from wireguard-go and can be surprisingly small; on Windows it is
+	// currently 2048 - 32 bytes.
+	maxTunMTU TunMTU = TunMTU(MaxPacketSize)
+	// A default "safe" MTU for the Tailscale TUN that we can use
+	// in the absence of other information such as path MTU probes.
+	safeTunMTU TunMTU = 1280
 )
 
-// DefaultMTU returns either the constant default MTU of 1280, or the value set
-// in TS_DEBUG_MTU clamped to a maximum of 65536.
-func DefaultMTU() uint32 {
-	// DefaultMTU is the Tailscale default MTU for now.
-	//
-	// wireguard-go defaults to 1420 bytes, which only works if the
-	// "outer" MTU is 1500 bytes. This breaks on DSL connections
-	// (typically 1492 MTU) and on GCE (1460 MTU?!).
-	//
-	// 1280 is the smallest MTU allowed for IPv6, which is a sensible
-	// "probably works everywhere" setting until we develop proper PMTU
-	// discovery.
-	tunMTU := defaultMTU
-	if mtu, ok := envknob.LookupUintSized("TS_DEBUG_MTU", 10, 32); ok {
-		mtu := uint32(mtu)
-		if mtu > maxMTU {
-			mtu = maxMTU
+// WireMTUsToProbe are the MTUs we are most likely to see in real world
+// networks. These are used to send discovery pings in packets padded
+// to exactly match the goal MTU.
+//
+// XXX These need adjusting before merging to main. What do you think?
```

Comment:
reminder to remove all "XXX" before merging. (we check in TODO but never "XXX" so you can always depend on XXX being greppable to find your own temporary comments)

Reply:
The XXX is here precisely so I don't merge it. :) Any thoughts on which MTU size to probe?

Reply:
No clue. I had always assumed the plan was to binary search it rather than try some specific values.

No opinions, as I don't really know what the design is.

Are these some sizes to spray out all at once to get a starting point to find the specific number? How often?

Reply:
Yeah, the observation is that in reality, there are only a small number of MTU values. Rather than binary search for them, just try them all at the same time and take the largest. The plan is to limit MTU discovery to once every 10 minutes, which is the standard for normal PMTUD cache expiration.

Reply:
There's a study on this subject, but it doesn't include jumbo frames since those can't usually transit the Internet
https://aura.abdn.ac.uk/bitstream/handle/2164/11693/tma2018_paper57.pdf
It... is surprisingly lacking in helpful answers for this question though.

This would be my list to guess from:

- Interface MTU if >9000
- 9000 (common jumbo frame choice)
- Some intermediate jumbo sizes. I've seen guides that recommend 8192 because it's a multiple of page size but that doesn't include the Ethernet header so I'm unclear it actually does what they expect. Perhaps 2048 - 14, 4096 - 14, 8192 - 14?
- 1500
- 1492 (PPPoE)
- 1476 (GRE)
- 1436? (IPsec ESP with GRE says maximum of 64 bytes overhead [e.g. Zscaler guide](https://help.zscaler.com/zia/determining-the-optimal-mtu-for-gre-or-ipsec-tunnels))
- 1400 (Zscaler guide recommends this as a pretty safe option for most cases)
- 1280 (IPv6 minimum, current Tailscale value)

If you want a super short list, I guess 9000, 1500, 1492, 1400, 1280 seems good?

Converging quickly on one of those and then probing a bit further to find the exact value later seems like an ideal approach, but is a little annoying because I don't think TCP is willing to increase the MSS after it's initially established. I believe it can be reduced later. It might be worth accepting the cost of some fragmentation at first before the value has been nailed down if TCP can be coaxed into decreasing it later.

RFC 1191 says that the next-hop MTU must be populated in the ICMP error, but RFC 792 does not, so fast probing a few options and then using that to guide finding an exact value (binary search between options otherwise) might work well. It's going to be impossible to get this perfect because the correct answer can change as peers switch between different networks from time to time (e.g. a laptop that moves around).

Reply:
Thanks! So the standard PMTUD algorithm actually recommends stopping binary search when it gets "close enough" rather than getting it to the exact bytes, and it discards the cached PMTU entry every 10 minutes and restarts discovery, so that's what people are living with right now. I like the idea of adding an underlying interface MTU as one of the probe values.

Reply:
How close is "close enough", without doing the math it surely doesn't save much more than two packets?

I think what I'd be tempted to do is to _not discard_ and to instead use the already discovered MTU for a lightweight retest, so test [cur, cur+1]. If cur+1 succeeds, then initiate a new "full probe" from [cur+1, max]. If both fail, initiate full probe with the default range.

Another value we might be able to gather on _some platforms_ as a starting point is to request the MSS from getsockopt TCP_INFO on platforms where it/equivalents are available, for any socket that's going through `netns`. This would give us a solid starting point.

Reply:
These are all great ideas! I am going to remove the probed MTUs array part of the patch and hard code the max probed MTU for the purposes of this PR for splitting up MTU concepts, which shouldn't include that stuff anyway, and then incorporate these ideas in the PR that actually probes the MTU. Thank you!